source/manuals/programming-languages/nodejs/index.html.md.erb

```diff
@@ -80,24 +117,321 @@ Keeping tests in the same directory makes them easier to find and more obvious w
+const userLoggedInCallback = function (authToken) { ... }
+const dbRequestResultCallback = function (req, res) { ... }
+```
+
+Separating out a callback may pose a problem when it needs its original scope. This can be handled by currying the callback:
+
+```
+fs.readFile(filename, callback(filename))
+
+const callback = filename => (err, data) =>
+  if (err) {
+    console.log(`failed reading ${filename}.`)
+    throw err
+  }
+  ...
+}
+```
+
+
+### Functional programming
+
+> Use JavaScript's functional programming features
+
+JavaScript has the advantage that it offers functional programming concepts
+natively, like functions as first-class objects, higher-order functions (like
+`map`, `reduce` or `apply`) or pattern matching.
+
+Following functional programming principles, such as immutable data structures
+and pure functions, produces code that is easier to test, less prone to runtime
+errors and is often more performant. Write functions as expressions that return
+values of a single type.
+
+Side effects in a function can have bad consequences, especially within `map` or
+`reduce`. `this` is better avoided, as it can refer to different objects
+depending on the context in which a function is executed and lead to unexpected
+side-effects.
+
+Remain aware that JavaScript is not just a functional language and you will
+most probably have to mix functional and object-oriented concepts, which can
+be tricky to get right.
+
+
+More guidance:
+
+- [An Introduction to Functional JavaScript](https://www.sitepoint.com/introduction-functional-javascript/)
+- [JavaScript Allongé](https://leanpub.com/javascriptallongesix/read)
+
+
+## Errors
+
+> Make sure you handle all errors
+
+Envisage all error scenarios, in particular in asynchronous callback functions
+or Promises, and have a fallback for any situation. Consider programmer errors
+(bugs) as well as operational errors (arising from external circumstances,
+like a missing file). Bugs that are caught by exceptions should be logged and
+the execution stopped, and the supervisor will restart the process. Use the
+built-in `Error` object instead of custom types. It makes logging easier.
+
+Your application shouldn't trust any input, for example from a file or an
+API. So the application should handle all operational errors and recover from
+them.
+
+## Node.js's HTTP server
+
+> Offload Node's server as much as possible
+
+Don't expose Node's HTTP server to the public. Use a reverse proxy to serve
+static assets, and cache content as much as possible. Implement adequate
+supervising: [pm2](http://pm2.keymetrics.io/) is recommended for Node.js-specific
+servers.
+
+## Transpiling
+
+> Stick to JavaScript
+
+Avoid anything that compiles to JavaScript (except for static type checking, see
+below). Examples to avoid include CoffeeScript, PureScript and [many others](https://github.com/jashkenas/coffeescript/wiki/List-of-languages-that-compile-to-JS).
+
+> Use TypeScript
```

Comment:
Maybe this should be changed to something along the lines of:

> If you require static typing consider using TypeScript. This has the advantage of... [list advantages]. But please be aware that there are disadvantages too... [list disadvantages]

Rather than a recommendation it could offer an opinion and list of considerations to be made on a project by project basis.

Reply:
I spoke to Max and he recognised that he's not had a chance to use Typescript in a project yet, we should just remove this for now until people have had a chance to try it out and see how it works.

Reply:
good points. I've modified the text accordingly.